src/index/base.cpp

```diff
@@ -105,41 +103,16 @@ bool BaseIndex::Init()
         SetBestBlockIndex(locator_index);
     }
 
-    // Skip pruning check if indexes are not ready to sync (because reindex-chainstate has wiped the chain).
-    const CBlockIndex* start_block = m_best_block_index.load();
-    bool synced = start_block == active_chain.Tip();
-    if (!synced && g_indexes_ready_to_sync) {
-        const CBlockIndex* block_to_test = start_block;
-        if (!block_to_test) block_to_test = active_chain.Genesis();
-
-        // Assert block_to_test is not null here. It can't be null because the
-        // genesis block can't be null here. The genesis block will be null
-        // during this BaseIndex::Init() call if the node is being started for
-        // the first time, or if -reindex is used. But in both of these cases
-        // m_best_block_index is also null so this branch is not reached.
-        assert(block_to_test);
-
-        if (!active_chain.Contains(block_to_test)) {
-            // if the bestblock is not part of the mainchain, find the fork
-            // so we can make sure we have all data down to the fork
-            block_to_test = active_chain.FindFork(block_to_test);
-        }
-
-        // make sure we have all block data back to the start block
-        if (!m_chainstate->m_blockman.CheckBlockDataAvailability(*active_chain.Tip(), *Assert(block_to_test))) {
-            return InitError(strprintf(Untranslated("%s best block of the index goes beyond pruned data. Please disable the index or reindex (which will download the whole blockchain again)"), GetName()));
-        }
-    }
-
     // Child init
+    const CBlockIndex* start_block = m_best_block_index.load();
```

Comment:
In commit "index: verify blocks data existence only once" (5ec5a4888743ce7261e0e9cdc077014cd47bfdd6)

Noticed this while I was rebasing #24230, but one side effect of this commit is now `CustomInit` will be called unnecessarily when the index can't start up because there is a pruning violation. This doesn't seem ideal, but it probably not worth worrying about. It would be good to mention the change in the commit description, though.

I think #24230 will restore previous behavior, and not call `CustomInit` when the index won't start. But it hard to avoid calling `CustomInit` here because it can't be moved to the Start function without also moving the `m_synced` assignment (because `CustomInit` has to be called before m_synced is true to make sure it is called before CustomAppend, as explained
3126454dcfa1dd29bb66500d5f2b5261684d6c58). Moving `m_synced` assignment to Start here would potentially be bad for performance, because it would mean indexes would only sync after the blocks were connected instead at the same time during LoadBlocks. #24230 should avoid calling CustomInit unnecessarily, and avoid needing to wait for LoadBlocks when the index is in sync by registering for notifications in a different way that doesn't rely on `m_synced` to avoid spurious notifications.

Reply:
> Noticed this while I was rebasing https://github.com/bitcoin/bitcoin/pull/24230, but one side effect of this commit is now CustomInit will be called unnecessarily when the index can't start up because there is a pruning violation. This doesn't seem ideal, but it probably not worth worrying about. It would be good to mention the change in the commit description, though.

Hmm, I hold the opposite view on this matter. I believe it is beneficial to call `CustomInit` within the base class's `Init` method, regardless of any external chain state. This approach ensures consistency in how an object is initialized, as it is isolated from its context initially.

By performing specific initialization checks on the index subclasses before anything else, we can guarantee data consistency. Moreover, the verification process in the child classes for `CustomInit` is faster, and any errors detected are more severe than the pruning violation. Currently, all failures within `CustomInit` are related to database corruption errors.